src/core/state.go

```diff
@@ -917,6 +917,77 @@ func (state *BuildState) waitForTargetAndEnsureDownload(l, dependent BuildLabel,
 	return target
 }
 
+// ActivateTarget marks a target as active (ie. to be built) and adds its dependencies as pending parses.
+func (state *BuildState) ActivateTarget(pkg *Package, label, dependent BuildLabel, forSubinclude bool) error {
+	if !label.IsAllTargets() && state.Graph.Target(label) == nil {
+		if label.Subrepo == "" && label.PackageName == "" && label.Name == dependent.Subrepo {
+			if subrepo := state.CheckArchSubrepo(label.Name); subrepo != nil {
+				state.ArchSubrepoInitialised(label)
+				return nil
+			}
+		}
+		if state.Config.Bazel.Compatibility && forSubinclude {
+			// Bazel allows some things that look like build targets but aren't - notably the syntax
+			// to load(). It suits us to treat that as though it is one, but we now have to
+			// implicitly make it available.
+			exportFile(state, pkg, label)
+		} else {
+			msg := fmt.Sprintf("Parsed build file %s but it doesn't contain target %s", pkg.Filename, label.Name)
+			if dependent != OriginalTarget {
+				msg += fmt.Sprintf(" (depended on by %s)", dependent)
+			}
+			return fmt.Errorf(msg + suggestTargets(pkg, label, dependent))
+		}
+	}
+	if state.ParsePackageOnly && !forSubinclude {
+		return nil // Some kinds of query don't need a full recursive parse.
+	} else if label.IsAllTargets() {
+		if dependent == OriginalTarget {
+			for _, target := range pkg.AllTargets() {
+				// Don't activate targets that were added in a post-build function; that causes a race condition
+				// between the post-build functions running and other things trying to activate them too early.
+				if state.ShouldInclude(target) && !target.AddedPostBuild {
+					// Must always do this for coverage because we need to calculate sources of
+					// non-test targets later on.
+					if !state.NeedTests || target.IsTest() || state.NeedCoverage {
+						if err := state.QueueTarget(target.Label, dependent, dependent.IsAllTargets()); err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	} else {
+		for _, l := range state.Graph.DependentTargets(dependent, label) {
+			// We use :all to indicate a dependency needed for parse.
+			if err := state.QueueTarget(l, dependent, forSubinclude || dependent.IsAllTargets()); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+
+// exportFile adds a single-file export target. This is primarily used for Bazel compat.
```

Comment:
Why do we need a special function here just for Bazel compat? Seems like the wrong place.

Reply:
Yeah, I agree but I don't have any better ideas at this time. Essentially we have a `load("//pkg:go.build_defs")` and load has to figure out if that's a real target or a file. Maybe we could queue the package up for parse, and if the target doesn't exist, we try adding this filegroup? Might make more sense in a followup PR though.